File: neovim-plugin/lua/cadence.lua
local M = {}

local config = {
  api_key = vim.env.CADENCE_API_KEY or "",
  endpoint = vim.env.CADENCE_ENDPOINT or "https://vs-integrate.vercel.app/api/heartbeat",
  debounce_ms = 30000,
  idle_timeout = 120,
}

local timer = nil
local last_activity = vim.loop.now()

local function json_encode(payload)
  return vim.fn.json_encode(payload)
end

local function current_project()
  return vim.fn.getcwd()
end

local function project_hash(project)
  return vim.fn.sha256(project):sub(1, 16)
end

local function post(payload)
  if config.api_key == "" then
    vim.notify("Cadence API key is not configured", vim.log.levels.WARN)
    return
  end

  local body = json_encode(payload)
  vim.fn.jobstart({
    "curl", "-sS", "-X", "POST",
    "-H", "Content-Type: application/json",
    "-H", "Authorization: Bearer " .. config.api_key,
    "-d", body,
    config.endpoint,
  }, { detach = true })
end

local function heartbeat(connection_test)
  local project = current_project()
  local file = vim.api.nvim_buf_get_name(0)
  local language = vim.bo.filetype ~= "" and vim.bo.filetype or "unknown"
  local now = os.time()
  local payload = {
    apiKey = config.api_key,
    ide = "neovim",
    timestamp = now * 1000,
    type = connection_test and "connection_test" or nil,
    language = language,
    file = file ~= "" and vim.fn.fnamemodify(file, ":e") or nil,
    project = vim.fn.fnamemodify(project, ":t"),
    projectHash = project_hash(project),
    platform = vim.loop.os_uname().sysname:lower(),
    isIdle = ((vim.loop.now() - last_activity) / 1000) > config.idle_timeout,
    timezoneOffset = -math.floor(os.difftime(os.time(os.date("!*t")), os.time(os.date("*t"))) / 60),
    localDate = os.date("%Y-%m-%d"),
  }
  post(payload)
end

local function queue()
  last_activity = vim.loop.now()
  if timer then
    timer:stop()
    timer:close()
  end
  timer = vim.loop.new_timer()
  timer:start(config.debounce_ms, 0, vim.schedule_wrap(function()
    heartbeat(false)
  end))
end

function M.setup(opts)
  config = vim.tbl_extend("force", config, opts or {})
  vim.api.nvim_create_user_command("CadenceTestConnection", function()
    heartbeat(true)
    vim.notify("Cadence connection test sent")
  end, {})

  vim.api.nvim_create_autocmd({ "BufEnter", "TextChanged", "TextChangedI", "BufWritePost" }, {
    group = vim.api.nvim_create_augroup("CadenceTelemetry", { clear = true }),
    callback = queue,
  })
end

return M

File: sublime-package/cadence.py
import hashlib
import json
import os
import platform
import threading
import time
import urllib.request

import sublime
import sublime_plugin


class CadenceState:
    timer = None
    last_activity = time.time()


def settings():
    return sublime.load_settings("Cadence.sublime-settings")


def project_hash(window):
    folders = window.folders() if window else []
    project = folders[0] if folders else ""
    return hashlib.sha256(project.encode("utf-8")).hexdigest()[:16] if project else ""


def send_heartbeat(view, connection_test=False):
    cfg = settings()
    api_key = cfg.get("api_key", "")
    endpoint = cfg.get("endpoint", "https://ca-dence.vercel.app/api/heartbeat")
    if not api_key:
        sublime.status_message("Cadence API key is not configured")
        return

    window = view.window() if view else sublime.active_window()
    file_name = view.file_name() if view else ""
    language = (view.settings().get("syntax") or "unknown").split("/")[-1].split(".")[0].lower() if view else "unknown"
    payload = {
        "apiKey": api_key,
        "ide": "sublime",
        "timestamp": int(time.time() * 1000),
        "type": "connection_test" if connection_test else None,
        "language": language,
        "file": os.path.splitext(file_name or "")[1].lstrip(".") or None,
        "project": os.path.basename(window.folders()[0]) if window and window.folders() else "unknown",
        "projectHash": project_hash(window),
        "platform": platform.system().lower(),
        "isIdle": (time.time() - CadenceState.last_activity) > cfg.get("idle_timeout", 120),
        "timezoneOffset": -int(time.localtime().tm_gmtoff / 60) if hasattr(time.localtime(), "tm_gmtoff") else 0,
        "localDate": time.strftime("%Y-%m-%d"),
    }
    body = json.dumps({k: v for k, v in payload.items() if v is not None}).encode("utf-8")
    request = urllib.request.Request(
        endpoint,
        data=body,
        headers={
            "Content-Type": "application/json",
            "Authorization": "Bearer " + api_key,
        },
        method="POST",
    )
    try:
        urllib.request.urlopen(request, timeout=5).read()
        sublime.status_message("Cadence heartbeat sent")
    except Exception as exc:
        sublime.status_message("Cadence heartbeat failed: {}".format(exc))


def queue_heartbeat(view):
    CadenceState.last_activity = time.time()
    if CadenceState.timer:
        CadenceState.timer.cancel()
    delay = settings().get("debounce_ms", 30000) / 1000
    CadenceState.timer = threading.Timer(delay, lambda: send_heartbeat(view))
    CadenceState.timer.daemon = True
    CadenceState.timer.start()


class CadenceEventListener(sublime_plugin.EventListener):
    def on_modified_async(self, view):
        queue_heartbeat(view)

    def on_post_save_async(self, view):
        queue_heartbeat(view)

    def on_activated_async(self, view):
        queue_heartbeat(view)


class CadenceTestConnectionCommand(sublime_plugin.ApplicationCommand):
    def run(self):
        view = sublime.active_window().active_view() if sublime.active_window() else None
        send_heartbeat(view, True)
